src/database/entities/server.rs

```diff
@@ -131,13 +130,81 @@ impl Server {
         Ok(())
     }
 
-    pub async fn publish_update(&self, data: Value) -> Result<()> {
-        ClientboundNotification::ServerUpdate {
-            id: self.id.clone(),
-            data,
-            clear: None,
+    pub async fn publish_update(&self, to_user: Option<&str>) -> Result<()> {
+        if let Some(user_id) = to_user {
+            let mut channels = Vec::new();
+
+            // Find the channels
+            let mut channel_cursor = get_collection("channels")
+                .find(
+                    doc! {
+                        "_id": {
+                            "$in": &self.channels
+                        }
+                    },
+                    None
+                )
+                .await
+                .map_err(|_| Error::DatabaseError {
+                    operation: "find",
+                    with: "channels",
+                })?;
+
+            // Only include channels the user can see
+            let user = User::from_id(user_id).await?;
+
+            while let Some(maybe_channel) = channel_cursor.next().await {
+                if let Ok(channel_document) = maybe_channel {
+                    let channel: Channel = from_document(channel_document)
+                        .map_err(|_| Error::DatabaseError {
+                            operation: "from_document",
+                            with: "channel",
+                        })?;
+
+                    let perms = PermissionCalculator::new(&user)
+                        .with_channel(&channel)
+                        .for_channel().await?;
+
+                    if perms.get_view() {
+                        channels.push(channel.id().into());
+                    }
+                }
+            }
```

Comment:
I forgot to document this anywhere, but my intention was to provide all channel IDs and require   clients to determine what they can see.